### ISSUE TITLE
loadPackage (recurse or otherwise) should respect __path__

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10867]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#1276|
|Type|defect|
|Created|2005-10-14 10:18:27Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1276 1276
type__defect defect
reporter__jml jml
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1129285107000000 1129285107000000
changetime__1153727293000000 1153727293000000
version__ 
owner__ 
cc__radix cc__jml cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
Implementation in trunk r14850 only handles __path__ for the top-level package.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
Do we still want this?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
I still want the repo reorganization, which requires this, yes. I was hoping 
someone else would pick up my initial solution for this issue (which no longer has 
the problem below) and run with it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Some points:
 * It's really hard to do this correctly
 * I don't want to create a new Trial-specific annotation variable for what feels like a Python problem
 * This issue covers a use case which is very rare

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

An alternative to the title would be to somehow inform trial of how to reverse map a given filename to a module.

Recap:
I want to move twisted into e.g. ```Core/twisted/internet/``` and ```Project/twisted/project/```). Then, make a ```twisted/__init__.py```, with a ```__path__``` that includes both ```Project/twisted``` and ```Core/twisted```.

SO, the problem really is: how do I run trial on all the code. Right now, you run "trial twisted". This does more than one might expect: it gets the file for the module "twisted", and does a recursive directory walk over all modules it can find within it. This recursive directory walk has a few problems: it disregards any __path__s which direct python to look other places for modules, and it only works on modules that are actually in the filesystem (rather than in a zip file).

Anyhow, a recursive directory walk, starting with "." would be okay in the scenario above, except that, after doing the walk, trial then has to make _backwards_ from filename to module. This is not really possible in a general way. So trial just assumes that the file "```foo/bar/baz.py```" is module "```foo.bar.baz```". This is often true, but not in the case above, where the file is named "```Project/twisted/project/foo.py```" and the module it contains is "```twisted.project.foo```".

So it would also solve this issue to have some way to tell trial "no, map the filenames back to module names like _this_"

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

I have a new opinion on how this should work:
* As before, we should walk everything, importing packages using __path__.
* Turn errors when trying to import a package to look for tests into some sort of "skipped" message. Don't bother with any extra custom attributes that might tell it not to look there/etc.
* Use the python 2.5 function pkgutils.walk_packages to do this, rather than implementing it ourselves. (and backport those functions to t.p.pymodules for use in prior python versions).

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Blaaah, that'll teach me not to use preview. stupid formatting. Reposting readably.

I have a new opinion on how this should work:

 * As before, we should walk everything, importing packages to look for their ```__path__```.
 * Turn errors when trying to import a package to look for tests into some sort of "skipped" message. Don't bother with any extra custom attributes that might tell it not to look there/etc.
 * Use the python 2.5 function pkgutils.walk_packages to do this, rather than implementing it ourselves. (and backport those functions to t.p.pymodules for use in prior python versions).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

I like this much more.

Wrt your second point: by "skipped", you mean that import errors should **not** fail the suite?

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

This idea:

ImportErrors, when raised while recursively trawling for packages to look for test modules in, will result in a "skipped" warning of some sort which should be relatively unobtrusive but still noticable. This is because it can be expected that this will happen, but it should still be reported on the off chance (unlikely) that it is preventing a package with tests from being found. 

ImportErrors, when raised trying to either run a test, or trying to load a "test_foo" module, will result in an error message as it does right now.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

OK. Sounds good (provided it's possible to tell the difference).

I'll look at this after [#1940](https://github.com/twisted/twisted/issues/1940) and [#1781](https://github.com/twisted/twisted/issues/1781) (and its dependants [#1936](https://github.com/twisted/twisted/issues/1936) and [#1883](https://github.com/twisted/twisted/issues/1883)) land.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

The branch for [#1940](https://github.com/twisted/twisted/issues/1940) fixes this as far as I understand it was supposed to be solved, i.e. without actually importing packages, because they might be things that aren't supposed to be imported.  It will honor ```__path__``` for already-imported packages, however, and and there is also a flag that can be passed to the one place ```walkModules``` is called in trial to import packages as they are discovered, makin the conversion to actually import vs. not be trivial to implement.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `closed`|
|-|-|

(In [17651]) Make it possible to run trial on tests located within a zip file.

Fixes [#1940](https://github.com/twisted/twisted/issues/1940)

Fixes [#10867](https://github.com/twisted/twisted/issues/10867)

Author: glyph

Reviewer: jml

This change also introduces a major new piece of functionality,
twisted.python.modules, as well as twisted.python.zippath, which allows for
FilePath-like access to zipfiles.

As part of changing trial to support zipfiles, another issue was discovered and
fixed, which is that trial's sort support was previously broken.  This is now
fixed, and trial will, by default, sort tests alphabetically now, by fully
qualified name.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

Reverted merge because of [#1952](https://github.com/twisted/twisted/issues/1952).  Re-opening.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `closed`|
|-|-|

(In [17672]) Re-merge branch for [[#1940](https://github.com/twisted/twisted/issues/1940)](https://github.com/twisted/twisted/issues/1940), plus fix for Windows and additional docs.

Fixes [#1952](https://github.com/twisted/twisted/issues/1952)
Fixes #1940
Fixes [#10867](https://github.com/twisted/twisted/issues/10867)

Author: glyph

Reviewer: jerub

This is a do-over of r17651, hopefully this time without the test failure on
the Windows buildslave.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
